MMF-3807: Raise awareness about recent changes and newly released versions

thahnen opened the pull request:
## Summary

Currently, Eclipse users don't know what has changed when updating SonarLint except for taking a look at the GitHub release page or the Sonar Community forum announcement. Additionally, they're not aware if a new version of SonarLint for Eclipse was released except if they trigger the **Check for Updates** action in Eclipse.
And even if users did not install it from our own Update Site (or the marketplace that links to it), they won't be informed about new versions.

We want to change that by raising awareness about what has changed by providing the Release Notes directly inside the IDE and informing the user after a fresh installation/update. For newer versions available, we check once per day against our own Update Site, and in case a newer version is available, we inform the user as well, encouraging them to update.

## Testing

This behavior only happens once the IDE loads, and if a newer version is available for the new version check, we cannot test that via integration tests for now. We can only check the new Release Notes preference page. Even the Welcome view we edited cannot be tested as this is a web browser and its "paging" cannot be interacted with using RedDeer.

thahnen committed:
SLE-863: Raise awareness about Release Notes in the IDE

Leverage the Eclipse UI "Welcome" page and our SonarLint integration to raise awareness about recent changes.
This is done about notifying the user after the installation / updating from an older version.

thahnen committed:
SLE-864: Inform about newer versions available

Every day SonarLint will check (if not disabled) whether a newer version was released in the meantime.
In that case, inform the user via a notification.

thahnen committed:
SLE-863: Integration test for the Release Notes preferences

Simple test to check that the Release Notes preferences is working correctly.

nquinquenel commented on org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/http/HttpClient.java:
```diff
@@ -0,0 +1,187 @@
+ *  Very bare bones implementation of a HTTP client that can work in the context of SonarLint both with a configurable
+ *  key and trust store. Additionally, timeouts can be provided that will used the same way as SLCORE would use them.
+ */
+public class HttpClient {
```
It seems like reinventing the wheel a lot. Couldn't you leverage the existing HTTP client in SLCORE?

thahnen replied:
Sadly no :cry:

Before 10.0, it was still available in the `org.sonarsource.sonarlint.core.clientapi.SonarLintBackend` but only to open HTTP connections to a specific connection. Since 10.0 it was completely moved to the part out of process, not being available to the clients anymore.

I thought about sparking a conversation with the Sonar Solution squad about it, but in the end decided against it for the following reasons:
- having an HTTP client as API would enable (malicious) third parties to plug into that and use it to trigger calls on behalf of SonarLint (Core), e.g. when SonarLint is the only plug-in configured with HTTP/SSL configurations and others are not
- in general, no client should make any HTTP requests as everything they need is either in SLCORE or provided by the IDE, here this is a solution as Eclipse doesn't provide what is needed (and probably never will be) - information about the Release Notes as well as newly released versions
- the implementation is as bare-bones as it gets, I was even thinking in the first place of dropping the `SSLContext` related logic but keeping it otherwise might have rendered this feature useless - the HTTP/SSL configuration provided by users is there for a reason in the first place

nquinquenel replied:
I was not thinking about using the HTTP client in SLE directly, but instead creating a dedicated service in SLCORE. This could be, perhaps, reused in other flavors.

thahnen commented on org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/SonarLintUiPlugin.java:
```diff
@@ -259,8 +265,38 @@ public IStatus run(IProgressMonitor monitor) {
+
+          if (newestSonarLintVersion == null) {
+            SonarLintLogger.get().debug("Cannot check for newer SonarLint versions via the SonarLint for Eclipse "
+              + "Update Site. In thase case, please check the Sonar Community Forum ("
```
Good catch, will fix that later

thahnen committed:
SLE-864: Inform users about new version

Display the new version and log both the current and new version for debugging.

nquinquenel reviewed (approved):
Not a huge fan of the HTTP service but let's go like that for now
